scripts/scons_helpers/build-3rdparty.py

```diff
@@ -664,18 +664,164 @@ def splitenv(envlist):
+    if 'linux' in toolchain and 'arm' in toolchain and bitness == 32:
+        return 'linux-generic32'
+
+    # android*: 'i[0-9]86' -> 'x86'
+    if ('android' in toolchain or 'linux' in toolchain) \
+        and '86' in toolchain and bitness == 32:
+        toolchain = re.sub(r'i\d86', 'x86', toolchain)
+
+    ### generic detection
+
+    # XXX: this is a dirty parser based on very simple heuristics and verified on a
+    # limited set of platforms
+
+    # order is significant here as we search substrings in a whole toolchain string
+    platforms = [
+        # [(os,          [arch_compiler, ...]), ...]
+        ('BSD',          ['aarch64', 'generic32', 'generic64', 'ia64', 'riscv64',
+                          'sparc64', 'sparcv8', 'x86-elf', 'x86_64', 'x86']),
+        ('Cygwin',       ['i386', 'i486', 'i586', 'i686', 'x86_64', 'x86']),
+        ('UEFI',         ['x86_64', 'x86']),
+        ('UWIN',         []),
+        ('aix64',        ['gcc-as', 'gcc', 'cc']),
+        ('aix',          ['gcc', 'cc']),
+        ('android64',    ['aarch64', 'mips64', 'x86_64']),
+        ('android',      ['arm64', 'armeabi', 'arm', 'mips64', 'mips', 'x86_64', 'x86']),
+        ('darwin64',     ['arm64-cc', 'arm64', 'ppc-cc', 'ppc', 'x86_64-cc', 'x86_64']),
+        ('darwin',       ['i386-cc', 'i386', 'ppc-cc', 'ppc']),
+        ('iossimulator', ['cross', 'xcrun']),
+        ('ios64',        ['cross', 'xcrun']),
+        ('ios',          ['cross', 'xcrun']),
+        ('linux32',      ['s390x']),
+        ('linux64',      ['loongarch64', 'mips64', 'riscv64', 's390x', 'sparcv9']),
+        ('linux',        ['aarch64', 'alpha-gcc', 'aout', 'arm64ilp32', 'armv4', 'c64xplus',
+                          'elf', 'generic32', 'generic64', 'ia64', 'latomic', 'mips32',
+                          'mips64', 'ppc64le', 'ppc64', 'ppc', 'sparcv8', 'sparcv9',
+                          'x86_64-clang', 'x86_64', 'x86-clang', 'x86', 'x32']),
+        ('mingw64',      []),
+        ('mingw',        []),
+    ]
+
+    #~print('0: ', toolchain)
+    for os, archs_compilers in platforms:
+        if os in toolchain:
+            for ac in archs_compilers:
+                #~print('1: ', os, ac)
+                # corner case for BSD
+                if ac == 'x86-elf':
+                    return '-'.join([os, ac])
+                # general case: 'arch-compiler'
+                ac_arr = ac.split('-')
+                arch = ac_arr[0]
+                if arch not in toolchain:
+                    continue
+                if len(ac_arr) > 1:
+                    compiler = ac_arr[1]
+                    if compiler in toolchain:
+                        return '-'.join([os, arch, compiler])
+                #~print('2: ', os, arch)
+                return '-'.join([os, arch])
+            if os in ['mingw', 'mingw64', 'Cygwin', 'UEFI', 'UWIN', 'android64']:
+                return os
+    if 'gcc' in toolchain:
+        return 'gcc'
```

Comment:
I think this can never happen. CMIIAW.

But you can use `checkfamily(env, toolchain, 'gcc')`.